### ISSUE TITLE
`rejectUnauthorized` option ignored in webpacked typescript (impacts mlxprs extension in latest VSCode on Mac)

### ISSUE THREAD

**Issue by mtcotton** [marklogic/node-client-api#568]
The `rejectUnauthorized` option is being ignored when `eval`ing against a node MarkLogic Database Client in some scenarios of webpacked typescript - in our case, the latest version of VS Code. I've put together a typescript test that I can debug and run in VScode to see the issue.

This worked in previous versions of VS Code, so my naive assumption is that there was a JS update that causes this break. I'm unfamiliar enough with typescript and webpack to pull the nest apart.

The use case: In order to execute queries against private DHS standard instances, we create SSH tunnels through an EC2 bastion host, and use mlxprs eval queries on the MarkLogic server. This is our primary replacement for query console in the field, and is critical to DHS projects.

The work-around: older versions (pre-March 2021) versions of VS Code appear to be unaffected, so as a work-around, we can revert to an older version.

The test:
```
import * as assert from 'assert';
import * as ml from 'marklogic';

// You can import and use all API from the 'vscode' module
// as well as import your extension to test it
import * as vscode from 'vscode';
// import * as myExtension from '../../extension';

suite('Extension Test Suite', () => {
	vscode.window.showInformationMessage('Start all tests.');

	test('MarkLogic acceptUnauthorized SSL Test', async () => {
		const query = `
            const options = {};
            if (modulesDb) { options.modules = xdmp.database(modulesDb) };
            if (contentDb) { options.database = xdmp.database(contentDb) };
            xdmp.eval(actualQuery,
                {actualQuery: actualQuery, sqlQuery: sqlQuery, sqlOptions: sqlOptions},
                options
                );`;
        const extVars = {
            "actualQuery": "'use strict';\n\n\"Success!\"",
            "contentDb": "data-hub-FINAL",
            "modulesDb": "data-hub-MODULES",
            "sqlQuery": "",
            "sqlOptions": []
        };
        const mldbc = ml.createDatabaseClient({
            "host": "localhost",
            "port": TUNNEL_PORT,
            "database": "data-hub-FINAL",
            "user": "MLUSERNAME",
            "password": "MLPASSWORD",
            "authType": "BASIC",
            "ssl": true,
            "ca": "",
            "rejectUnauthorized": false
        });
		const result = await mldbc.eval(query, extVars).result()
		assert.ok(true, JSON.stringify(result));
	})
});
```

The command to create SSH tunnel:
```
ssh -L TUNNEL_PORT:MLHOST:8011 -i PRIVATE_KEY_PATH ec2-user@BASTION_PUBLIC_IP
```

**Comment by ehennum** [comment]
@mtcotton, has this issue isolated to the MarkLogic Node.js API?  

If the root cause is ambiguous and the fix is critical for DHS, it may be appropriate to engage DHS support.

If the issue is with mlxprs, please file the bug in the mlxprs GitHub repository.  

I ask because -- while it is always possible that there's an unexplained issue -- a quick look at the source code for the Node.js API shows that rejectUnauthorized is passed through to the Node.js Agent in the same way as all of the other SSL properties.

    https://github.com/marklogic/node-client-api/blob/master/lib/marklogic.js#L767-L771

In other words, the Node.js API isn't doing anything special with rejectUnauthorized, so if any of the SSL properties are passing through correctly to the Agent, the bug isn't likely to be in the Node.js API.

(Pinging @anu3990 for awareness.)

**Comment by ehennum** [comment]
Recording the resolution from an offline discussion ...

Node.js distinguishes between the HTTP request options and the agent options with extra options for both when using SSL.

The refactoring to use the native Node.js Agent introduced a howling gaffe by setting the SSL extra HTTP request options as Agent options.

The initClient() function in the marklogic.js module should change as follows...

The isSSL flag should be set at the top of the function instead of here:

https://github.com/marklogic/node-client-api/blob/master/lib/marklogic.js#L757-L759

If the isSSL flag is true, the keys array here:

https://github.com/marklogic/node-client-api/blob/master/lib/marklogic.js#L711

should add these configuration keys:  ca, cert, ciphers, clientCertEngine, crl, dhparam, ecdhCurve, honorCipherOrder, key, passphrase, pfx, rejectUnauthorized, secureOptions, secureProtocol, servername, sessionIdContext, highWaterMark

The agent options without SSL set here:
 
https://github.com/marklogic/node-client-api/blob/master/lib/marklogic.js#L780

should process these options:  keepAliveMsecs, maxSockets, maxTotalSockets, maxFreeSockets, scheduling, timeout

The agent options with SSL set here:

https://github.com/marklogic/node-client-api/blob/master/lib/marklogic.js#L767

should process these options and also this extra option: maxCachedSessions

Many thanks to @mtcotton for identifying and isolating the issue.

**Comment by anu3990** [comment]
@mtcotton the develop branch now has the fix. Please let us know if it works. Thank you.!

**Comment by mtcotton** [comment]
> @mtcotton the develop branch now has the fix. Please let us know if it works. Thank you.!

Looks good to me! I'll forward this to the mlxprs ticket.

**Comment by georgeajit** [comment]
Shipping the issue.
